drivers/hmis/app/graphql/types/hmis_schema/enrollment.rb

```diff
@@ -449,5 +450,57 @@ def intake_assessment
     def exit_assessment
       load_ar_association(object, :exit_assessment)
     end
+
+    def last_contact
+      last_contact_entity = [
+        # Service DateProvided may be nil, so to avoid n+1 queries, sort with nulls_first using arel
+        # (rather than adding .where clause to the scope, which won't work - see comments on load_ar_association).
+        load_ar_association(
+          object,
+          :services,
+          scope: Hmis::Hud::Service.order(Hmis::Hud::Service.arel_table[:date_provided].asc.nulls_first),
+        ).last,
+        # CustomService DateProvided is guaranteed non-null by DB
+        load_ar_association(object, :custom_services, scope: Hmis::Hud::CustomService.order(:date_provided)).last,
+        # CLS InformationDate is guaranteed non-null by DB
+        load_ar_association(object, :current_living_situations, scope: Hmis::Hud::CurrentLivingSituation.order(:information_date)).last,
+        # AssessmentDate is guaranteed non-null by DB
+        load_ar_association(object, :custom_assessments, scope: Hmis::Hud::CustomAssessment.order(:assessment_date)).last,
+        # CustomCaseNote's information_date can be null in DB, so sort with nulls_first using arel
+        load_ar_association(
+          object,
+          :custom_case_notes,
+          scope: Hmis::Hud::CustomCaseNote.order(Hmis::Hud::CustomCaseNote.arel_table[:information_date].asc.nulls_first),
+        ).last,
+      ].compact.
+        max_by { |entity| Hmis::Hud::Enrollment.contact_date_for_entity(entity) } # max_by treats nils as lower than any non-nil value
```

Comment:
I'm seeing this raise `comparison of NilClass with Date failed` if any entities have a missing date field. 
Is there a reason not to just omit records with null values from the scopes above?

Reply:
Hm, so, maybe I am misunderstanding the comments on [load_ar_association](https://github.com/greenriver/hmis-warehouse/blob/6c41a9d74b937594b6dc868833cb06afd927f680/drivers/hmis/app/graphql/concerns/graphql_application_helper.rb#L35-L38):
> Note: 'scope' is intended for ordering or to modify the default association in a way that is constant with respect to the resolver, for example `scope: FooBar.order(:name)`. It is NOT used to filter down results.

This is the reason I avoided adding a `.where` clause that omits null values from the scopes. But I tried it just now and it does seem to work.

Reply:
Ok, I updated it https://github.com/greenriver/hmis-warehouse/pull/5153/commits/cccb7094822dc88dac8fa6d71b1b13707eee9b7a and it does make the code a lot more sensible and readable. I got rid of all the confusing comments that were trying to make up for the lack of clarity. It passed the tests, including the n+1, so I guess it should be functional. 

I still feel a bit nervous about it though, since the load_ar_association so explicitly states that the "scope" passed should not be used to filter down results...

Reply:
Good point, sorry for the confusion! That comment should be updated. The "modify the default association in a way that is constant with respect to the resolver" part is accurate. Maybe changing the end to:
> It should not be used to filter down results, _unless_ the filter is constant with respect to the resolver.
> For example: `FooBar.where.not(foo: nil).order(:name)` is ok, but `FooBar.where(name: object.name)` is not


Does that make sense? It just can't be dependent on the resolver at all.